src/live_restore/live_restore_fs.c

```diff
@@ -608,29 +639,34 @@ __live_restore_fh_write(
     lr_fh = (WT_LIVE_RESTORE_FILE_HANDLE *)fh;
     session = (WT_SESSION_IMPL *)wt_session;
 
+    WT_ASSERT_ALWAYS(session, __wt_rwlock_islocked(session, &lr_fh->ext_lock),
+      "Live restore lock not taken when needed");
     __wt_verbose_debug1(
       session, WT_VERB_FILEOPS, "WRITE %s: %" PRId64 ", %lu", fh->name, offset, len);
+
     WT_RET(lr_fh->destination.fh->fh_write(lr_fh->destination.fh, wt_session, offset, len, buf));
-    WT_RET(lr_fh->destination.fh->fh_sync(lr_fh->destination.fh, wt_session));
-    WT_RET(__live_restore_remove_extlist_hole(lr_fh, session, offset, len));
-    return (0);
+    return (__live_restore_remove_extlist_hole(lr_fh, session, offset, len));
 }
 
 /*
- * __read_promote --
- *     Write out the contents of a read into the destination. This will be overkill for cases where
- *     a read is performed to service a write.
+ * __live_restore_fh_write --
+ *     File write.
  */
 static int
-__read_promote(WT_LIVE_RESTORE_FILE_HANDLE *lr_fh, WT_SESSION_IMPL *session, wt_off_t offset,
-  size_t len, char *read)
+__live_restore_fh_write(
+  WT_FILE_HANDLE *fh, WT_SESSION *wt_session, wt_off_t offset, size_t len, const void *buf)
 {
-    __wt_verbose_debug2(session, WT_VERB_FILEOPS, "    READ PROMOTE %s : %" PRId64 ", %lu",
-      lr_fh->iface.name, offset, len);
-    WT_RET(
-      __live_restore_fh_write((WT_FILE_HANDLE *)lr_fh, (WT_SESSION *)session, offset, len, read));
+    WT_DECL_RET;
+    WT_LIVE_RESTORE_FILE_HANDLE *lr_fh;
+    WT_SESSION_IMPL *session;
 
-    return (0);
+    lr_fh = (WT_LIVE_RESTORE_FILE_HANDLE *)fh;
+    session = (WT_SESSION_IMPL *)wt_session;
+
+    __wt_writelock(session, &lr_fh->ext_lock);
+    ret = __live_restore_fh_write_int(fh, wt_session, offset, len, buf);
+    __wt_writeunlock(session, &lr_fh->ext_lock);
```

Comment:
I think this would be cleaner and less error prone to write locking macros similar to the (many) other ones in `schema.h`. You could look at `WT_WITH_HANDLE_LIST_READ|WRITE_LOCK` and `WT_WITH_TABLE_READ|WRITE_LOCK`. Some have `NOWAIT` variations. 

Instead of setting a session lock flag, you'd set one on the specific `lr_fh` instead. But it nicely encapsulates usage. Nearly all the uses of your new lock would fit very well into that style.

Reply:
I certainly see the case for WRITE lock but none of the readlock usages match that pattern, I'll go ahead and implement both macros in any case.

Reply:
See: 375f2c302